toolkit/tools/pkg/imagecustomizerlib/customizeuki.go

```diff
@@ -0,0 +1,476 @@
+		}
+		if !initramfsExists {
+			return nil, fmt.Errorf("missing initramfs for kernel: (%s), expected (%s)", kernelName, initramfsName)
+		}
+
+		// Add to the map if both exist.
+		kernelToInitramfs[kernelName] = initramfsName
+	}
+
+	return kernelToInitramfs, nil
+}
+
+func createUki(uki *imagecustomizerapi.Uki, buildDir string, buildImageFile string) error {
+	logger.Log.Debugf("Customizing UKI")
+
+	var err error
+
+	loopback, err := safeloopback.NewLoopback(buildImageFile)
+	if err != nil {
+		return fmt.Errorf("failed to connect to image file to provision UKI:\n%w", err)
+	}
+	defer loopback.Close()
+
+	diskPartitions, err := diskutils.GetDiskPartitions(loopback.DevicePath())
+	if err != nil {
+		return err
+	}
+
+	systemBootPartition, err := findSystemBootPartition(diskPartitions)
+	if err != nil {
+		return err
+	}
+	bootPartition, err := findBootPartitionFromEsp(systemBootPartition, diskPartitions, buildDir)
+	if err != nil {
+		return err
+	}
+
+	systemBootPartitionTmpDir := filepath.Join(buildDir, tmpEspParitionDirName)
+	systemBootPartitionMount, err := safemount.NewMount(systemBootPartition.Path, systemBootPartitionTmpDir, systemBootPartition.FileSystemType, 0, "", true)
+	if err != nil {
+		return fmt.Errorf("failed to mount esp partition (%s):\n%w", bootPartition.Path, err)
+	}
+	defer systemBootPartitionMount.Close()
+
+	bootPartitionTmpDir := filepath.Join(buildDir, tmpParitionDirName)
+	bootPartitionMount, err := safemount.NewMount(bootPartition.Path, bootPartitionTmpDir, bootPartition.FileSystemType, 0, "", true)
+	if err != nil {
+		return fmt.Errorf("failed to mount partition (%s):\n%w", bootPartition.Path, err)
+	}
+	defer bootPartitionMount.Close()
+
+	osSubreleaseFullPath := filepath.Join(buildDir, UkiBuildDir, "os-release")
+	stubPath := filepath.Join(buildDir, UkiBuildDir, "linuxx64.efi.stub")
+	cmdlineFilePath := filepath.Join(buildDir, UkiBuildDir, "kernel-cmdline-args")
+
+	// Get mapped kernels and initramfs.
+	kernelToInitramfs, err := getKernelToInitramfsMap(bootPartitionTmpDir, uki.Kernels)
+	if err != nil {
+		return err
+	}
+
+	for kernel, initramfs := range kernelToInitramfs {
+		err := buildUki(kernel, initramfs, cmdlineFilePath, osSubreleaseFullPath, stubPath, buildDir,
+			systemBootPartitionTmpDir,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to build UKI for kernel (%s):\n%w", kernel, err)
+		}
+	}
+
+	err = cleanupUkiBuildDir(buildDir)
+	if err != nil {
+		return fmt.Errorf("Error during cleanup UKI build dir:\n%w", err)
+	}
+
+	err = cleanupBootPartition(bootPartitionTmpDir)
+	if err != nil {
+		return fmt.Errorf("failed to clean up boot partition:\n%w", err)
+	}
+
+	err = systemBootPartitionMount.CleanClose()
+	if err != nil {
+		return err
+	}
+
+	err = bootPartitionMount.CleanClose()
+	if err != nil {
+		return err
+	}
+
+	err = loopback.CleanClose()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func extractKernelCmdlineFromGrub(grubCfgPath string) (string, error) {
```

Comment:
this can be moved to `grubcfgutils.go` where we keep all processing of grub.cfg.

Reply:
see `FindLinuxLine`

Reply:
Thanks George, the `extractKernelToArgsFromGrub` is dedicated to UKI for now, though I agree it can be a general function. Let's deprioritize this change, and implement when developing bootloader structure.